### PR TITLE
zephyr: boot_serial fixes

### PR DESCRIPTION
**Pull request by carlescufi**
Multiple fixes for boot_serial in Zephyr based on feedback from the Zephyr kernel experts.

Signed-off-by: Carles Cufi <carles.cufi@nordicsemi.no>

### CODE REVIEW

File: boot/zephyr/serial_adapter.c [hunk 1]
```diff
@@ -53,138 +53,157 @@ static int boot_uart_fifo_init(void);
 int
 console_out(int c)
 {
-    uart_poll_out(uart_dev, c);
+	printk("tx: %d\n", c);
```
Comment (mkiiskila) [inline boot/zephyr/serial_adapter.c]: Is printk()'s ok here and in console_read()?

**Comment by nvlsianpu** [comment]
I observed situation when `used_queue` was inexhaustible - will debug this at Monday, so please hold one the PR a little.

**Comment by nvlsianpu** [comment]
@carlescufi ^^

File: boot/zephyr/serial_adapter.c (hunk 1 above)
Comment (carlescufi) [reply boot/zephyr/serial_adapter.c]: Nope, thanks for catching this. This was a mistake. I have removed this now

File: boot/zephyr/serial_adapter.c [hunk 2]
```diff
@@ -53,138 +53,154 @@ static int boot_uart_fifo_init(void);
...
+		return (-1);
+	}
 
-    if (!uart_dev) {
-        return (-1);
-    }
+	/* Enable all interrupts unconditionally. Note that this is due
+	 * to Zephyr issue #8393. This could should be removed once the
+	 * issue is fixed in upstream Zephyr. */
+	irq_unlock(0);
```
Comment (carlescufi) [inline boot/zephyr/serial_adapter.c]: @cvinayak can you please review this line?

Comment (cvinayak) [reply boot/zephyr/serial_adapter.c]: Any reason interrupts are enabled before the callback is setup? It is better to enable at function return.

**Comment by carlescufi** [comment]
@nvlsianpu cherry-picked your commits here

**Comment by carlescufi** [comment]
@mkiiskila @utzig could you please re-review?
@utzig do you know why CI is failing?

**Comment by carlescufi** [comment]
@cvinayak fixed

**Comment by utzig** [comment]
@carlescufi You seem to have cherry-picked commit 95a60036d3e577811b577c3724d84cbae5710a13, so you need to add your SoB along with Andrzej's.

**Comment by carlescufi** [comment]
@utzig thanks! fixed now

**Comment by carlescufi** [comment]
@utzig ok this is ready. It fixes the serial issues completely and it also substantially reduces the size of the binary when building with Zephyr

File: boot/boot_serial/src/boot_serial.c [hunk 3]
```diff
@@ -586,16 +582,21 @@ boot_serial_start(const struct boot_uart_funcs *f)
...
-          buf[1] == SHELL_NLIP_DATA_START2) {
-            rc = boot_serial_in_dec(&buf[2], off - 2, dec, &dec_off, max_input);
+            rc = boot_serial_in_dec(&in_buf[2], off - 2, dec_buf, &dec_off, max_input);
+        } else if (in_buf[0] == SHELL_NLIP_DATA_START1 &&
+          in_buf[1] == SHELL_NLIP_DATA_START2) {
+            rc = boot_serial_in_dec(&in_buf[2], off - 2, dec_buf, &dec_off, max_input);
+        }
+
+        /* serve errors: out of decode memory, or bad encoding */
+        if (rc == -1) {
```
Comment (utzig) [inline boot/boot_serial/src/boot_serial.c]: if this is not a diff issue, one empty block was left behind here (I see the correct handling just below)

Comment (carlescufi) [reply boot/boot_serial/src/boot_serial.c]: @nvlsianpu can you please reply since this is in one of your commits?

Comment (nvlsianpu) [reply boot/boot_serial/src/boot_serial.c]: Rightm, This if can be used for error in protocol parsing signaling, as it is empty - should be removed.

**Comment by mkiiskila** [comment]
Looks ok.

File: boot/boot_serial/src/boot_serial.c (hunk 3 above)
Comment (nvlsianpu) [reply boot/boot_serial/src/boot_serial.c]: @carlescufi It Is fixed on my fork.

Comment (carlescufi) [reply boot/boot_serial/src/boot_serial.c]: @utzig Fixed now

**Pull request merged**